Create Wizard pattern (#5647)
jpmorganchase/salt-ds: site/docs/patterns/wizard.mdx

description: "The wizard allows a workflow-based task to be displayed in a sequence of manageable steps, for easier completion."
layout: DetailPattern
tags:
  - pattern
aliases:
  - /salt/patterns/wizard
data:
  resources:
    [
      { href: "https://storybook.saltdesignsystem.com/?path=/story/patterns-wizard", label: "Examples"},
    ]
  components: ["Button Bar", "Dialog", "Stepper"]
---

<Diagram
  src="/img/patterns/wizard/wizard-pattern-example.png"
  srcDark="/img/patterns/wizard/wizard-pattern-example-dark.png"
  alt="Example of a multi-step wizard workflow with stepper navigation"
  border
  background="secondary"
/>

## When to use

Use wizard to guide users through a series of steps or tasks in a specific, linear order, often with the goal of completing a complex process (e.g., software installation, account setting/onboarding, checkout). Users must complete each step in order before moving to the next.

## How to build

### Anatomy

Wizard contains:
1. **Workflow title**: use a standardized, actionable title that clearly reflects the intent of the wizard (e.g., “Create a new account”). The title should remain consistent across all steps to maintain context.

> 2. **Step title**: to reflect the current step which vary per step. Step titles should be concise, descriptive, and clearly reflect the purpose of each step. Use plain language that is easy for users to understand.

Reviewer: ```suggestion
2. **Step title**: to reflect the current step. Step titles should be concise, descriptive, and clearly reflect the purpose of each step. Use plain language that's easy for users to understand.
```

### Wizard with a horizontal stepper

<Diagram
  src="/img/patterns/wizard/wizard-pattern-layout-horizontal.png"
  srcDark="/img/patterns/wizard/wizard-pattern-layout-horizontal-dark.png"
  alt="Horizontal wizard layout with stepper and content area spacing"
  border
  background="secondary"
/>

### Wizard with a vertical stepper

<Diagram
  src="/img/patterns/wizard/wizard-pattern-layout-vertical.png"
  srcDark="/img/patterns/wizard/wizard-pattern-layout-vertical-dark.png"
  alt="Vertical wizard layout with stepper and content area arrangement"
  border
  background="secondary"
/>

<GuidanceCallout type="positive" customPillText="Best practices">
  The wizard should maintain a consistent size as users navigate between steps, ensuring a stable and predictable layout throughout the workflow.
</GuidanceCallout>

## Horizontal stepper vs vertical stepper

We recommend using a horizontal stepper for shorter and simpler tasks. For more lengthy and complex task flows, use a vertical stepper to break down the process into steps and substeps.

### The number of steps

> For simpler flows, a horizontal stepper with fewer steps is usually sufficient. If your process involves a large amount of content, we recommend a vertical stepper and a wide, shallow structure. This means breaking the process into more steps with fewer fields per step (use Miller's Law of 7 ± 2 items as a guide) which are easier for users to complete and recollect. Additionally, use substeps to further group closely related fields. This approach can reduce cognitive load and improves usability compared to having fewer steps with many fields.

Reviewer: ```suggestion
For simpler flows, a horizontal stepper with fewer steps is usually sufficient. If your process involves a large amount of content, we recommend a vertical stepper and a wide, shallow structure. This means breaking the process into more steps with fewer fields per step (use Miller's Law of 7 ± 2 items as a guide) which are easier for users to complete and recollect. Additionally, use substeps to further group closely related fields. This approach can reduce cognitive load and improve usability compared to having fewer steps with many fields.
```

</GuidanceCallout>

## Modal wizard

You can use a dialog component as a structure base for a modal wizard with a stepper when you need to insert the wizard within another task or flow. The header in the dialog should include the wizard title and stepper. 

For shorter tasks that temporarily become the user’s main focus, a modal experience is preferable, as it prevents interaction with the rest of the UI and helps focus the user’s attention on the specific task or information at hand.

<Diagram
  src="/img/patterns/wizard/wizard-pattern-modal.png"
  srcDark="/img/patterns/wizard/wizard-pattern-modal-dark.png"
  alt="Modal wizard example with stepper and dialog container"
  border
  background="secondary"
/>

<GuidanceCallout type="positive" customPillText="Best practices">
Do not include a close button, as it can be ambiguous whether the progress or the entered information will be saved or lost. Use a cancel button instead to clearly indicate the user’s intent to exit the flow, ensuring a conscious, deliberate action to stop the process.
</GuidanceCallout>

## Optional step

> When a step is entirely optional and contains no required fields, you can indicate this beneath the step title (in stepper). We also recommend to use the “All fields are optional” text in the description under the step title.

Reviewer: ```suggestion
When a step is entirely optional and contains no required fields, you can indicate this beneath the step title (in the stepper). We also recommend to use the “All fields are optional” text in the description under the step title.
```

  srcDark="/img/patterns/wizard/wizard-pattern-optional-step-dark.png"
  alt="Wizard step with header description indicating all fields are optional"
  border
  background="secondary"
  caption="“All fields are optional” text in the description under the step title"
/>

<GuidanceCallout type="positive" customPillText="Best practices">
  Indicate the minority: When a step contains both required and optional fields, label only the few optional or required fields as needed.
</GuidanceCallout>

## Validation state

Wizard can show validation feedback. Feedback can occur either in real time (dynamically) as the user interacts with the form or after submission (when clicking the final action button, e.g., Submit), depending on the validation approach implemented.

When validation feedback is triggered:
- The stepper updates to reflect the validation state.
- All action buttons remain enabled throughout the flow for accessibility and discoverability.
- You should give the user options to resolve issues before they proceed. 
- Users can still go back or cancel the flow if needed.

<GuidanceCallout type="positive" customPillText="Best practices">

>   When navigating back through previous steps, the current step is always updated to show ‘Active’ and the steps ahead of the users current step should always be shown as ‘Pending’ (this includes any steps which were showing validation errors, since these may be corrected, or become unnecessary as the user progresses through the wizard).

Reviewer: ```suggestion
  When navigating back through previous steps, the current step is always updated to show ‘Active’ and the steps ahead of the user's current step should always be shown as ‘Pending’ (this includes any steps which were showing validation errors, since these may be corrected, or become unnecessary as the user progresses through the wizard).
```

</GuidanceCallout>

### Warning

If a step contains a warning after validation:
- The step displays a warning icon.
- The next button remains enabled, as the next step can be accessed without the issue being resolved.
- The warning can appear dynamically (real-time validation) or after the action (on submission), depending on the system’s validation logic.
- Users can go back to review or adjust information as needed.

<Diagram
  src="/img/patterns/wizard/wizard-validation-warning.png"
  srcDark="/img/patterns/wizard/wizard-validation-warning-dark.png"
  alt="Wizard stepper displaying a warning validation state after step validation"
  border
  background="secondary"
  caption="The step in the stepper displays a warning state."
/>

### Error

When a step contains an error after validation:
- The step displays an error state

> - The application highlights the issues in the specific fields or areas that need attention to be resolved.

Reviewer: ```suggestion
- Issues in the specific fields or areas that need attention to be resolved are highlighted.
```

- The next or submit button remains enabled for accessibility and discoverability, but clicking it again should refocus or re-highlight unresolved errors.
- The cancel and previous buttons remain enabled, allowing users to exit or navigate backward in the flow.

> - If the validation passes, the flow proceeds to the next step or confirmation screen.

Reviewer: This point is a bit weird, "When a step contains an error after validation, if the validation passes..."

Author: Updated to: "If all errors are resolved and validation passes, the user proceeds to the next step or confirmation screen."

<Diagram
  src="/img/patterns/wizard/wizard-validation-error.png"
  srcDark="/img/patterns/wizard/wizard-validation-error-dark.png"
  alt="Wizard stepper displaying an error validation state"
  border
  background="secondary"
  caption="The step in the stepper displays an error state"
/>

## Confirmation messages

In certain use cases, you may want to provide additional feedback to users after an action is triggered. This includes showing warning, success, or other confirmation messages that help users stay informed and confident in their next steps.

Confirmation messages can appear:
- as a dialog over the app’s main content area, or
- as part of the wizard flow within a modal wizard.

### Warning message

> When the user clicks the cancel button, display a warning message if there’s a risk of losing progress. This is especially useful in lengthy or critical flows where saving is not supported. After confirming the cancel action, it should restore the state to how it was before the wizard began.

Reviewer: ```suggestion
When the user clicks the cancel button, display a warning message if there’s a risk of losing progress. This is especially useful in lengthy or critical flows where saving isn't supported. After confirming the cancel action, it should restore the state to how it was before the wizard began.
```

#### Wizard with the warning dialog

<Diagram
  src="/img/patterns/wizard/wizard-confirmation-embedded-warning.png"
  srcDark="/img/patterns/wizard/wizard-confirmation-embedded-warning-dark.png"
  alt="Embedded warning dialog in wizard confirming cancel action and risk of losing progress"
  border
  background="secondary"
  caption="The warning message shows up to confirm if cancelling."
/>

#### Modal wizard with the warning message

<Diagram
  src="/img/patterns/wizard/wizard-confirmation-modal-warning.png"
  srcDark="/img/patterns/wizard/wizard-confirmation-modal-warning-dark.png"
  alt="Modal wizard displaying warning message to confirm cancel action"
  border
  background="secondary"
  caption="The warning message shows up to confirm if cancelling."
/>

## Completion message

> Use a completion message to confirm that the wizard task has been successfully completed. This helps users clearly understand that their action was successful and provides closure to the flow, especially when the task is critical. 

Reviewer: We don't need to change this but what if the task wasn't successfully completed?
